imap/guesstz.c

```diff
@@ -0,0 +1,914 @@
+        goto done;
+    }
+    buf_reset(&key);
+
+    /* Initialize db */
+    gtzdb = xzmalloc(sizeof(struct guesstzdb));
+    gtzdb->fname = xstrdup(fname);
+    gtzdb->db = db;
+
+    /* Version */
+    gtzdb->version = *data;
+    data++;
+    datalen--;
+
+    /* Timestamp of creation */
+    gtzdb->created = *((int64_t*)(data));
+    data += 8; datalen -= 8;
+
+    /* Start and end of time range */
+    int64_t tstart = *((int64_t*)(data));
+    int64_t tend = *((int64_t*)(data + 8));
+    data += 16; datalen -= 16;
+    gtzdb->dbstart = icaltime_from_timet_with_zone(tstart, 0, utc);
+    gtzdb->dbend = icaltime_from_timet_with_zone(tend, 0, utc);
+
+    /* Zero-terminated IANA version */
+    gtzdb->ianaversion = xstrdup(data);
+
+done:
+    if (r && db) {
+        cyrusdb_close(db);
+    }
+    buf_free(&key);
+    free(myfname);
+    return gtzdb;
+}
+
+EXPORTED void guesstz_close(struct guesstzdb **gtzdbptr)
+{
+    if (!gtzdbptr || !*gtzdbptr) return;
+
+    struct guesstzdb *gtzdb = *gtzdbptr;
+
+    int r = cyrusdb_close(gtzdb->db);
+    if (r) {
+        xsyslog(LOG_ERR, "cyrusdb_close", "err=<%s>", cyrusdb_strerror(r));
+    }
+    free(gtzdb->ianaversion);
+    free(gtzdb->fname);
+    buf_free(&gtzdb->buf1);
+    buf_free(&gtzdb->buf2);
+    free(gtzdb);
+
+    *gtzdbptr = NULL;
+}
+
+EXPORTED void guesstz_toiana(struct guesstzdb *gtzdb,
+                             struct buf *idbuf, icalcomponent *vtz,
+                             struct icalperiodtype span,
+                             unsigned is_recurring)
+{
+    const icaltimezone *utc = icaltimezone_get_utc_timezone();
+    dynarray_t obsrecs;
+    dynarray_init(&obsrecs, sizeof(struct obsrec));
+    struct buf key = BUF_INITIALIZER;
+
+    static strarray_t preferred = STRARRAY_INITIALIZER;
+    if (!preferred.count) {
+        /* A hand-curated list of timezone names that follow
+         * the typical pattern of shifting daylight savings
+         * time one hour from standard time. One for (almost)
+         * every offset hour */
+        strarray_append(&preferred, "US/Aleutian");
+        strarray_append(&preferred, "US/Alaska");
+        strarray_append(&preferred, "US/Pacific");
+        strarray_append(&preferred, "US/Mountain");
+        strarray_append(&preferred, "US/Central");
+        strarray_append(&preferred, "US/Eastern");
+        strarray_append(&preferred, "America/Puerto_Rico");
+        strarray_append(&preferred, "America/Nuuk");
+        strarray_append(&preferred, "Atlantic/Azores");
+        strarray_append(&preferred, "Europe/London");
+        strarray_append(&preferred, "Europe/Berlin");
+        strarray_append(&preferred, "Europe/Athens");
+        strarray_append(&preferred, "Indian/Mauritius");
+        strarray_append(&preferred, "Asia/Dhaka");
+        strarray_append(&preferred, "Australia/Melbourne");
+        strarray_append(&preferred, "Pacific/Norfolk");
+    }
```

Comment:
What's the purpose of this list?  Is this just used to select one of multiple matches for a given offset?

Reply:
Yes, most likely multiple timezones will match, and we prefer these names.